[quant] Regsiter fake_quant and observer attributes as buffers

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33626 [quant] Regsiter fake_quant and observer attributes as buffers**

Summary: For DDP we require the attributes to be registered as buffers. By doing this the value is broadcast from one device to the rest.

Test Plan:
Tested on actual model on GPU

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20038839](https://our.internmc.facebook.com/intern/diff/D20038839)

supriyar committed:
[quant] Regsiter fake_quant and observer attributes as buffers

Summary: For DDP we require the attributes to be registered as buffers. By doing this the value is broadcast from one device to the rest.

Test Plan:
Tested on actual model on GPU

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Regsiter fake_quant and observer attributes as buffers"


Summary: For DDP we require the attributes to be registered as buffers. By doing this the value is broadcast from one device to the rest.

Test Plan:
Tested on actual model on GPU

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20038839](https://our.internmc.facebook.com/intern/diff/D20038839)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Regsiter fake_quant and observer attributes as buffers"


Summary: For DDP we require the attributes to be registered as buffers. By doing this the value is broadcast from one device to the rest.

Test Plan:
Tested on actual model on GPU

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20038839](https://our.internmc.facebook.com/intern/diff/D20038839)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Regsiter fake_quant and observer attributes as buffers"


Summary: For DDP we require the attributes to be registered as buffers. By doing this the value is broadcast from one device to the rest.

Test Plan:
Tested on actual model on GPU

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20038839](https://our.internmc.facebook.com/intern/diff/D20038839)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/fake_quantize.py:
```diff
@@ -111,7 +111,8 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
             key = prefix + name
             if key in state_dict:
-                setattr(self, name, state_dict.pop(key))
+                val = state_dict[key]
```
Can we remove this function?

supriyar replied:
Removing this function throws an error which says that the  the size of the loaded tensor does not match the original size  i.e., These buffers start out with numel 0 and become numel 1 once they have their first forward pass.

jerryzh168 replied:
yeah, we can't, I have tried this before as well. Could you add a comment here? @supriyar

supriyar commented:
Link issue #33648

supriyar committed:
Update on "[quant] Regsiter fake_quant and observer attributes as buffers"


Summary: For DDP we require the attributes to be registered as buffers. By doing this the value is broadcast from one device to the rest.

Test Plan:
Tested on actual model on GPU

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20038839](https://our.internmc.facebook.com/intern/diff/D20038839)

[ghstack-poisoned]
